File: src/com/prey/activities/FeedBackActivity.java
package com.prey.activities;

import java.util.Calendar;
import java.util.Date;

import com.prey.PreyConfig;
import com.prey.R;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.net.Uri;
import android.os.Bundle;

public class FeedbackActivity extends PreyActivity {

	private static final int SHOW_POPUP = 0;

	public static int FLAG_FEEDBACK_INIT = 0;
	public static int FLAG_FEEDBACK_C2DM = 1;
	public static int FLAG_FEEDBACK_SEND = 2;
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
		showDialog(SHOW_POPUP);
	}

	@Override
	protected Dialog onCreateDialog(int id) {

		Dialog popup = null;
		switch (id) {

		case SHOW_POPUP:
			AlertDialog.Builder alert = new AlertDialog.Builder(this);
			alert.setIcon(R.drawable.logo);
			alert.setTitle(R.string.feedback_principal_title);
			alert.setMessage(R.string.feedback_principal_message);

			alert.setPositiveButton(R.string.feedback_principal_button1, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int id) {
					Intent intent = new Intent(Intent.ACTION_VIEW);
					intent.setData(Uri.parse("market://details?id=com.prey"));
					startActivity(intent);
					PreyConfig.getPreyConfig(getApplicationContext()).setFlagFeedback(FLAG_FEEDBACK_SEND);
					finish();
				}
			});

			alert.setNeutralButton(R.string.feedback_principal_button2, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int id) {
					Intent popup = new Intent(getApplicationContext(), FormFeedbackActivity.class);
					popup.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
					startActivity(popup);
					PreyConfig.getPreyConfig(getApplicationContext()).setFlagFeedback(FLAG_FEEDBACK_SEND);
					finish();
				}
			});

			alert.setNegativeButton(R.string.feedback_principal_button3, new DialogInterface.OnClickListener() {
				public void onClick(DialogInterface dialog, int id) {
					PreyConfig.getPreyConfig(getApplicationContext()).setFlagFeedback(FLAG_FEEDBACK_SEND);
					finish();
				}
			});
			popup = alert.create();
		}
		return popup;
	}

	public static boolean showFeedback(long installationDate, int flagFeedback) {
		if (flagFeedback == FLAG_FEEDBACK_C2DM) {
			return true;
		} else {
			if (flagFeedback == FLAG_FEEDBACK_INIT) {
				Calendar cal = Calendar.getInstance();
				cal.setTimeInMillis(installationDate);
				cal.add(Calendar.WEEK_OF_YEAR, 2);
				long twoWeekOfYear = cal.getTimeInMillis();
				long now = new Date().getTime();
				if (now > twoWeekOfYear) {
					return true;
				}
			}
		}
		return false;
	}
}

File: app/src/main/java/com/prey/activities/FeedbackActivity.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities

import android.app.AlertDialog
import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle

import com.prey.R
import com.prey.PreyConfig

import java.util.Calendar
import java.util.Date

/**
 * Activity responsible for displaying a feedback dialog to the user.
 */
class FeedbackActivity : PreyActivity() {

    /**
     * Called when the activity is created.
     *
     * @param savedInstanceState The saved instance state, or null if not saved.
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        showDialog(SHOW_POPUP)
    }

    /**
     * Creates a dialog for the given ID.
     *
     * @param id The ID of the dialog to create.
     * @return The created dialog.
     */
    override fun onCreateDialog(id: Int): Dialog {
        var dialog: Dialog? = null
        when (id) {
            SHOW_POPUP -> {
                val builder = AlertDialog.Builder(this)
                builder.setIcon(R.drawable.info)
                builder.setTitle(R.string.feedback_principal_title)
                builder.setMessage(R.string.feedback_principal_message)
                builder.setPositiveButton(
                    R.string.feedback_principal_button1
                ) { _, _ ->
                    val intent = Intent(Intent.ACTION_VIEW)
                    intent.setData(Uri.parse("market://details?id=com.prey"))
                    startActivity(intent)
                    PreyConfig.getInstance(applicationContext)
                        .setFlagFeedback(FLAG_FEEDBACK_SEND)
                    finish()
                }
                builder.setNeutralButton(
                    R.string.feedback_principal_button2
                ) { _, _ ->
                    val popup = Intent(applicationContext, FormFeedbackActivity::class.java)
                    popup.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                    startActivity(popup)
                    PreyConfig.getInstance(applicationContext)
                        .setFlagFeedback(FLAG_FEEDBACK_SEND)
                    finish()
                }
                builder.setNegativeButton(
                    R.string.feedback_principal_button3
                ) { _, _ ->
                    PreyConfig.getInstance(applicationContext)
                        .setFlagFeedback(FLAG_FEEDBACK_SEND)
                    finish()
                }
                dialog = builder.create()
                PreyConfig.getInstance(applicationContext).setActivityView(ACTIVITY_FEEDBACK)
            }
        }
        return dialog!!
    }

    /**
     * Checks if the feedback should be shown based on the installation date and feedback flag.
     *
     * @param installationDate The date the app was installed.
     * @param feedbackFlag The feedback flag.
     * @return True if the feedback should be shown, false otherwise.
     */
    fun showFeedback(installationDate: Long, feedbackFlag: Int): Boolean {
        return when (feedbackFlag) {
            FLAG_FEEDBACK_C2DM -> true
            FLAG_FEEDBACK_INIT -> {
                val calendar = Calendar.getInstance()
                calendar.timeInMillis = installationDate
                calendar.add(Calendar.WEEK_OF_YEAR, 2)
                val twoWeeksLater = calendar.timeInMillis
                val now = Date().time
                now > twoWeeksLater
            }

            else -> false
        }
    }

    companion object {
        private const val SHOW_POPUP = 0
        private const val FLAG_FEEDBACK_INIT = 0
        private const val FLAG_FEEDBACK_C2DM = 1
        private const val FLAG_FEEDBACK_SEND = 2
        const val ACTIVITY_FEEDBACK: String = "ACTIVITY_FEEDBACK"
    }

}